[WIP] python custom op API

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120345

Looking for API feedback on custom_ops3.py

[no-ci]

zou3519 committed:
[WIP] python custom op API

Following the proposal over at
https://dev-discuss.pytorch.org/t/rfc-new-python-operator-registration-api/1838.

Looking for initial API feedback (see custom_ops.py).

[ghstack-poisoned]

ezyang commented on custom_ops.py:
```diff
@@ -0,0 +1,78 @@
+    # the abstract impl. Must be "traceable". User must use opcheck to test.
+    @staticmethod
+    def abstract(x):
+        return torch.empty_like(x)
```
This is not really a "new API" problem, but do I get fake tensors or meta tensors as inputs to abstract?

ezyang replied:
(In particular, am I able to sniff the device on the tensors in order to vary striding layout on different devices?)

zou3519 replied:
You get FakeTensors if we're doing Fake propagation and meta tensors if we're doing operations with meta tensors. So yes.

I've been wondering if we need separate per-backend abstracts (e.g. abstract_cpu, abstract_cuda, ...), since there was an issue where a user wanted a way to add their own custom device and specify a custom abstract for it.

zou3519 replied:
They would need to call .build() again after doing the above

zou3519 replied:
we can get rid of build if the user doesn't change stuff

ezyang commented on custom_ops.py:
```diff
@@ -0,0 +1,78 @@
+    # autograd: provide us setup_backward() and backward() methods.
+    # these must be "traceable". User must use opcheck to test.
+    @staticmethod
+    def setup_backward(ctx, inputs, output):
```
Would it be too cute to name this "forward" lol. Probably.....

ezyang replied:
Is this only ever saving for backward? Would it be crazy to name this `save_for_backward`?

ezyang commented on custom_ops.py:
```diff
@@ -0,0 +1,78 @@
+
+    Returns the sin of x.
+    """
+    return my_sin_op(x)
```
Blehhhh boilerplate. Why not pass the docstring into build?

zou3519 replied:
It's just a couple more lines and it seems more natural this way. If we shove everything into build, then:
- does the user need to pass the type signature as well? (I guess we could generate it)
- I want to ban kwarg-only arguments in schemas for custom ops created with this API. Having the wrapper allows users to specify kwarg-only arguments.

ezyang commented on custom_ops.py:
```diff
@@ -0,0 +1,78 @@
+
+    @staticmethod
+    @traceable  # specifies that we can auto-generate an abstract impl for this op
+    @device_types("CPU", "CUDA")  # specifies which devices this is valid for
```
If it's traceable, is it ever necessary to limit this for a subset of devices? Seems suspicious.

zou3519 replied:
It's possible that it's traceable for only those two devices if the user has something like the following:
```
if x.is_cuda:
    return x.sin()
if x.is_cpu:
    return x.sin()
return weird_device_kernel_sin(x)
```
In general, CompositeExplicitAutograd and CompositeImplicitAutograd have really bitten us in the custom ops work. Users will write a custom op for CPU in mind, use one of those two keys, and then something bad happens when someone else tries to use the custom op on CUDA.

ezyang commented on custom_ops.py:
```diff
@@ -0,0 +1,78 @@
+# Builds the provided implementations into a "custom op" that behaves like a function.
+# We don't do something like MySin.apply (i.e. the autograd.Function way) because
+# dynamic changes to methods on the MySin object will not be included in the op.
+my_sin_op = MySin.build()
```
So, is this the *real* reason you did it this way? Or was it because you need to run some extra code after the class def. An alternate implementation strategy that doesn't require build() is a custom metaclass for OpDef, which performs the initialization work after the class is done being constructed.

It's not clear to me that the justification here for build holds much water. You could always just make sure you getattr(MySin, ...) inside of the registered dispatcher functions so you always are using the up to date function. Or if you close over MySin, this will happen automatically.

zou3519 replied:
This is the real reason why I did it this way, yes. I considered the metaclass approach. That seemed a bit too magical. I tried to explain this (the metaclass thing) to some folks once and it blew their minds.

I'm not completely opposed to it though

zou3519 replied:
> It's not clear to me that the justification here for build holds much water. You could always just make sure you getattr(MySin, ...) inside of the registered dispatcher functions so you always are using the up to date function. Or if you close over MySin, this will happen automatically.

So the problem is that build() is potentially expensive. For example, if someone updates their MySin object with a `@traceable(impl)`, then we need to figure that out in all the rules we generate from it (e.g. abstract impl, vmap rule). It seemed more explicit to have a build() function (the model is that the user provides implementations and then builds a custom op out of them).

albanD commented on library.py:
```diff
@@ -0,0 +1,174 @@
+
+def traceable(f):
+    @functools.wraps(f)
+    def wrapper(*args, **kwargs):
```
Why a wrapper and not directly on f? Static method shenanigans? :p

zou3519 replied:
yes

albanD commented on library.py:
```diff
@@ -0,0 +1,174 @@
+    impl_cuda_method = getattr(opdef, "impl_cuda", None)
+    if impl_cuda_method is not None:
+        register_backend(lib, name, impl_cuda_method, "CUDA")
+
```
I guess in the final version we'll want to register error kernels for all other backends and features)?

zou3519 committed:
Update on "[WIP] python custom op API"


Following the proposal over at
https://dev-discuss.pytorch.org/t/rfc-new-python-operator-registration-api/1838.

Looking for initial API feedback (see custom_ops.py). The implementation (library.py) is not hardened at all; opened this PR for API discussion.

[ghstack-poisoned]

zou3519 commented:
Updated the API with feedback:
- there's no build() anymore (we are going to assume the user isn't going to hot-patch these objects -- this is fairly rare in practice)
- user wrapping the code in a function is optional (like how it is for autograd.Function)
- removed `@device_types` annotation for impl (it was confusing)

albanD commented on library.py:
```diff
@@ -0,0 +1,207 @@
+            return result
+        opoverload = register(result)
+        result._opoverload = opoverload
+        return result
+
```
```suggestion
        return result
        
    def __setattr__(cls, name, val):
        super().__setattr__(name, val)
        # Add any field that should not trigger a re-register
        if name not in ["_opoverload"]:
            cls._opoverload = register(result)
```

If you want any update of the class to re-trigger a register, you can do this so that `MySin.impl_cpu = foo` will work.

zou3519 replied:
:O

zou3519 commented on custom_ops.py:
```diff
@@ -0,0 +1,76 @@
+
+    Returns the sin of x.
+    """
+    return MySin.call(x)
```
nit: apply? (From Alban)

zou3519 committed:
Update on "[WIP] python custom op API"


Looking for initial API feedback. There are actually two different APIs
in this file.
1. custom_ops.py: single unified `Operator` object
2. custom_ops2.py: split API (`BackendImpl` and `Function`)

Related: https://dev-discuss.pytorch.org/t/rfc-new-python-operator-registration-api/1838.

[no-ci]

[ghstack-poisoned]

zou3519 commented:
Updates:

There is now an alternative API that I'm trialing out.
- the original API is in custom_ops.py
- the alternative API is in custom_ops2.py

zou3519 commented on custom_ops2.py:
```diff
@@ -0,0 +1,208 @@
+#
+# This is intentionally confusingly named like torch.autograd.Function,
+# because it is the thing you should use instead of torch.autograd.Function :).
+class MySin(Function):
+    @staticmethod
+    def forward(x):
+        return NumpySin.apply(x)
+
+    @staticmethod
+    def post_forward(ctx, args, output):
+        x, = args
+        ctx.save_for_backward(x)
+
+    @staticmethod
+    def backward(ctx, grad_out):
+        return grad_out * ctx.saved_tensors[0].cos()
```
cc @tugsbayasgalan it's plausible that we could just take an autograd.Function and turn it into a custom op, with or without Dynamo, under the assumption that it is make_fx-traceable (and well behaved). The `.apply` on this automatically creates 1+ opoverloads depending on what the input types are.

zou3519 committed:
Update on "[WIP] python custom op API"


Looking for API feedback on custom_ops3.py

[no-ci]

[ghstack-poisoned]

zou3519 committed:
Update on "[WIP] python custom op API"


Looking for API feedback on custom_ops3.py

[no-ci]

[ghstack-poisoned]
